File: openapi/openapi-generator/src/main/java/ru/tinkoff/kora/openapi/generator/javagen/ClientRequestMapperGenerator.java
package ru.tinkoff.kora.openapi.generator.javagen;

import com.palantir.javapoet.*;
import org.openapitools.codegen.CodegenOperation;
import org.openapitools.codegen.CodegenParameter;
import org.openapitools.codegen.model.OperationsMap;

import javax.lang.model.element.Modifier;
import java.util.ArrayList;
import java.util.Objects;

import static ru.tinkoff.kora.openapi.generator.KoraCodegen.isContentJson;

public class ClientRequestMapperGenerator extends AbstractJavaGenerator<OperationsMap> {
    @Override
    public JavaFile generate(OperationsMap ctx) {
        var className = ClassName.get(apiPackage, ctx.get("classname") + "ClientRequestMappers");
        var b = TypeSpec.interfaceBuilder(className)
            .addAnnotation(generated());
        for (var operation : ctx.getOperations().getOperation()) {
            if (!operation.getHasFormParams()) {
                continue;
            }
            b.addType(buildFormMapper(ctx, className, operation));
        }

        return JavaFile.builder(apiPackage, b.build()).build();
    }

    private TypeSpec buildFormMapper(OperationsMap ctx, ClassName rootName, CodegenOperation operation) {
        var className = rootName.nestedClass(capitalize(operation.operationId) + "FormParamRequestMapper");
        var formParamClassName = ClassName.get(apiPackage, ctx.get("classname").toString(), capitalize(operation.operationId) + "FormParam");
        var b = TypeSpec.classBuilder(className)
            .addAnnotation(generated())
            .addModifiers(Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL)
            .addSuperinterface(ParameterizedTypeName.get(Classes.httpClientRequestMapper, formParamClassName));
        var constructor = MethodSpec.constructorBuilder()
            .addModifiers(Modifier.PUBLIC);
        var apply = MethodSpec.methodBuilder("apply")
            .returns(Classes.httpBodyOutput)
            .addParameter(Classes.context, "ctx")
            .addParameter(formParamClassName, "value")
            .addException(Exception.class)
            .addModifiers(Modifier.PUBLIC)
            .addAnnotation(Override.class);
        for (var p : operation.formParams) {
            if (requiresMapper(p)) {
                var mapperType = ParameterizedTypeName.get(Classes.stringParameterConverter, asType(p));
                constructor.addParameter(mapperType, p.paramName + "Converter")
                    .addStatement("this.$N = $N", p.paramName + "Converter", p.paramName + "Converter");
                b.addField(mapperType, p.paramName + "Converter", Modifier.PRIVATE, Modifier.FINAL);
            }
        }
        var urlEncodedForm = operation.consumes != null && operation.consumes.stream()
            .map(m -> m.get("mediaType"))
            .anyMatch("application/x-www-form-urlencoded"::equalsIgnoreCase);
        var multipartForm = operation.consumes != null && operation.consumes.stream()
            .map(m -> m.get("mediaType"))
            .anyMatch("multipart/form-data"::equalsIgnoreCase);
        if (urlEncodedForm && multipartForm) {
            throw new IllegalArgumentException("Unsupported form type: " + operation);
        }
        if (urlEncodedForm) {
            apply.addStatement("var b = new $T()", ClassName.get("ru.tinkoff.kora.http.client.common.form", "UrlEncodedWriter"));
            for (var formParam : operation.formParams) {
                apply.beginControlFlow("if (value.$N() != null)", formParam.paramName);
                if (requiresMapper(formParam)) {
                    apply.addStatement("b.add($S, $N.convert(value.$N()))", formParam.baseName, formParam.paramName + "Converter", formParam.paramName);
                } else {
                    apply.addStatement("b.add($S, $T.toString(value.$N()))", formParam.baseName, ClassName.get(Objects.class), formParam.paramName);
                }
                apply.endControlFlow();
            }
            apply.addStatement("return b.write()");
        } else if (multipartForm) {
            apply.addStatement("var l = new $T<$T>()", ClassName.get(ArrayList.class), Classes.formPart);
            for (var formParam : operation.formParams) {
                apply.beginControlFlow("if (value.$N() != null)", formParam.paramName);
                if (formParam.isFile) {
                    apply.addStatement("l.add(value.$N())", formParam.paramName);
                } else if (requiresMapper(formParam)) {
                    apply.addStatement("l.add($T.data($S, $N.convert(value.$N())))", Classes.formMultipart, formParam.baseName, formParam.paramName + "Converter", formParam.paramName);
                } else {
                    apply.addStatement("l.add($T.data($S, $T.toString(value.$N())))", Classes.formMultipart, formParam.baseName, ClassName.get(Objects.class), formParam.paramName);
                }
                apply.endControlFlow();
            }
            apply.addStatement("return $T.write(l)", Classes.multipartWriter);
        } else {
            throw new IllegalStateException();
        }

        return b.addMethod(constructor.build()).addMethod(apply.build()).build();
    }

    private boolean requiresMapper(CodegenParameter p) {
        if (isContentJson(p)) {
            return true;
        }
        if (p.isEnum || (p.allowableValues != null && !p.allowableValues.isEmpty())) {
            return true;
        }
        return !p.isPrimitiveType;
    }
}

File: openapi/openapi-generator/src/main/kotlin/ru/tinkoff/kora/openapi/generator/kotlingen/ClientRequestMapperGenerator.kt
package ru.tinkoff.kora.openapi.generator.kotlingen

import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import org.openapitools.codegen.CodegenOperation
import org.openapitools.codegen.CodegenParameter
import org.openapitools.codegen.model.OperationsMap
import ru.tinkoff.kora.openapi.generator.KoraCodegen.isContentJson


class ClientRequestMapperGenerator : AbstractKotlinGenerator<OperationsMap>() {
    override fun generate(ctx: OperationsMap): FileSpec {
        val className = ClassName(apiPackage, ctx["classname"].toString() + "ClientRequestMappers")
        val b = TypeSpec.interfaceBuilder(className)
            .addAnnotation(generated())
        for (operation in ctx.operations.operation) {
            if (operation.hasFormParams) {
                b.addType(buildFormMapper(ctx, className, operation))
            }
        }

        return FileSpec.get(apiPackage, b.build())
    }

    private fun buildFormMapper(ctx: OperationsMap, rootName: ClassName, operation: CodegenOperation): TypeSpec {
        val className = rootName.nestedClass(capitalize(operation.operationId) + "FormParamRequestMapper")
        val formParamClassName = ClassName(apiPackage, ctx["classname"].toString(), capitalize(operation.operationId) + "FormParam")
        val b = TypeSpec.classBuilder(className)
            .addAnnotation(generated())
            .addSuperinterface(Classes.httpClientRequestMapper.asKt().parameterizedBy(formParamClassName))
        val constructor = FunSpec.constructorBuilder()
        val apply = FunSpec.builder("apply")
            .returns(Classes.httpBodyOutput.asKt())
            .addParameter("ctx", Classes.context.asKt())
            .addParameter("value", formParamClassName)
            .addModifiers(KModifier.OVERRIDE)
        for (p in operation.formParams) {
            if (requiresMapper(p)) {
                val mapperType = Classes.stringParameterConverter.asKt().parameterizedBy(asType(p).asKt())
                val mapperName = p.paramName + "Converter"
                constructor.addParameter(mapperName, mapperType)
                b.addProperty(PropertySpec.builder(mapperName, mapperType).initializer(mapperName).build())
            }
        }
        val urlEncodedForm = operation.consumes != null && operation.consumes.asSequence()
            .map { m -> m["mediaType"] }
            .any { anotherString: String? -> "application/x-www-form-urlencoded".equals(anotherString, ignoreCase = true) }
        val multipartForm = operation.consumes != null && operation.consumes.asSequence()
            .map { m -> m["mediaType"] }
            .any { anotherString: String? -> "multipart/form-data".equals(anotherString, ignoreCase = true) }
        if (urlEncodedForm && multipartForm) {
            throw IllegalArgumentException("Unsupported form type: $operation")
        }
        if (urlEncodedForm) {
            apply.addStatement("val b = %T()", ClassName("ru.tinkoff.kora.http.client.common.form", "UrlEncodedWriter"))
            for (formParam in operation.formParams) {
                if (formParam.required) {
                    apply.beginControlFlow("value.%N.let", formParam.paramName)
                } else {
                    apply.beginControlFlow("value.%N?.let", formParam.paramName)
                }
                if (requiresMapper(formParam)) {
                    apply.addStatement("b.add(%S, %N.convert(it))", formParam.baseName, formParam.paramName + "Converter")
                } else {
                    apply.addStatement("b.add(%S, it.toString())", formParam.baseName)
                }
                apply.endControlFlow()
            }
            apply.addStatement("return b.write()")
        } else if (multipartForm) {
            apply.addStatement("val l = arrayListOf<%T>()", Classes.formPart.asKt())
            for (formParam in operation.formParams) {
                if (formParam.required) {
                    apply.beginControlFlow("value.%N.let", formParam.paramName)
                } else {
                    apply.beginControlFlow("value.%N?.let", formParam.paramName)
                }
                if (formParam.isFile) {
                    apply.addStatement("l.add(it)")
                } else if (requiresMapper(formParam)) {
                    apply.addStatement("l.add(%T.data(%S, %N.convert(it)))", Classes.formMultipart.asKt(), formParam.baseName, formParam.paramName + "Converter")
                } else {
                    apply.addStatement("l.add(%T.data(%S, it.toString()))", Classes.formMultipart.asKt(), formParam.baseName)
                }
                apply.endControlFlow()
            }
            apply.addStatement("return %T.write(l)", Classes.multipartWriter.asKt())
        } else {
            throw IllegalArgumentException("Unsupported form type: $operation")
        }

        b.addFunction(apply.build())
        b.primaryConstructor(constructor.build())
        return b.build()
    }

    private fun requiresMapper(p: CodegenParameter): Boolean {
        if (isContentJson(p)) {
            return true
        }
        if (p.isEnum || !p.allowableValues.isNullOrEmpty()) {
            return true
        }
        return !p.isPrimitiveType
    }

}
